### ISSUE TITLE
[x86] improve cost/codegen of insertelement of FP element 0

### ISSUE THREAD

**Issue by rotateright** [llvm/llvm-project#58261]
Based on an example from: https://reviews.llvm.org/D135278

Inserting an FP scalar into element 0 of an FP vector is assumed to be 0 cost, but that may not be accurate when the base vector is not undef:

```ll
define <4 x float> @ins0_v4f32(<4 x float> %v, float %s) {
  %r = insertelement <4 x float> %v, float %s, i32 0
  ret <4 x float> %r
}
```

```console
% opt -mtriple=x86_64 -passes="print<cost-model>" -disable-output -mattr=+sse2 inselt.ll
Printing analysis 'Cost Model Analysis' for function 'ins0_v4f32':
Cost Model: Found an estimated cost of 0 for instruction:   %r = insertelement <4 x float> %v, float %s, i32 0

% opt -mtriple=x86_64 -passes="print<cost-model>" -disable-output -mattr=+avx2 inselt.ll
Printing analysis 'Cost Model Analysis' for function 'ins0_v4f32':
Cost Model: Found an estimated cost of 0 for instruction:   %r = insertelement <4 x float> %v, float %s, i32 0

% llc -o - inselt.ll -mattr=sse2
	movss	%xmm1, %xmm0                    ## xmm0 = xmm1[0],xmm0[1,2,3]

% llc -o - inselt.ll -mattr=avx2
	vblendps	$1, %xmm1, %xmm0, %xmm0         ## xmm0 = xmm1[0],xmm0[1,2,3]
```

There's a comment in X86InstrSSE.td about this choice:
```
// Prefer a movss or movsd over a blendps when optimizing for size. these were
// changed to use blends because blends have better throughput on sandybridge
// and haswell, but movs[s/d] are 1-2 byte shorter instructions.
```
I'm not sure if that makes sense anymore. Either way, we probably want the cost model for insertelement to check if the base vector is undef or not?

**Comment by RKSimon** [comment]
This is going to have to wait until we can update the getVectorInstrCost() call - it currently only takes opcode/type/index arguments, but we need an actual (optional) Value for the extractfrom/insertinto vector argument as well. Ideally we'd add CostKind while we're there as well....

**Comment by RKSimon** [comment]
> This is going to have to wait until we can update the getVectorInstrCost() call - it currently only takes opcode/type/index arguments, but we need an actual (optional) Value for the extractfrom/insertinto vector argument as well. Ideally we'd add CostKind while we're there as well....

These were added at 9b5f62685ab447ba9d3ea8ac2616e0c76a44d21b and 5fb3a57ea759

**Comment by RKSimon** [comment]
Fixed by 0e1312fbe07c0c14b0317ef62163326b086e0e58
